### ISSUE TITLE
Fetching paths from cache.nixos.org can get 'stuck'

### ISSUE THREAD

**Issue by raboof** [NixOS/nixpkgs#160289]
### Describe the bug

Sometimes, nix gets stuck fetching paths:

```
$ nix-shell -p sbt
these 2 paths will be fetched (408.08 MiB download, 694.33 MiB unpacked):
  /nix/store/0kg8zm25k6av8hgf177kwnp9q5j6yrf3-openjdk-17.0.1+12
  /nix/store/wj7x8ik896h0k9r6fk0bns21xw5c18yh-sbt-1.6.2
copying path '/nix/store/0kg8zm25k6av8hgf177kwnp9q5j6yrf3-openjdk-17.0.1+12' from 'https://cache.nixos.org'...
```

Stalling without any network traffic.

Interestingly this seems to be somehow tied to the specific path: nix can download other paths just fine, but always gets stuck on (in this case) `/nix/store/0kg8zm25k6av8hgf177kwnp9q5j6yrf3-openjdk-17.0.1+12`. 

It has something to do with the system state: rebooting the machine makes it possible to fetch the path again, but killing all `nix-daemon` processes without rebooting doesn't seem to help.

edit: it does seem to be something around nix-daemon though, because when rebooting I see systemd-shutdown waiting for several nix-daemon processes to shut down which takes quite a while.

If anyone can suggest ways to further diagnose the problem I'd be happy to try things out!

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.16.5, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.6.0`
 - channels(root): `"nixos-20.03pre194293.2436c27541b"`
 - nixpkgs: `/home/aengelen/nixpkgs`
```

(current profile is at 9f697d60e4d, nixpkgs is at 48d63e924a)

**Comment by kenranunderscore** [comment]
I'm seeing the same just now. For me lots of builds worked today, but I get stuck at `/nix/store/062izp7vafcszxbip77mwya90rzkcg19-mbrola-3.3` specifically, also after killing the nix-daemon processes.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.95, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.7.0pre20220127_558c4ee`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by 06kellyjac** [comment]
It seems to be related to remote builds. I have to `sudo kill -9` the processes

https://discourse.nixos.org/t/remote-building-experienence-massively-degraded-since-2-3-months/16950

**Comment by raboof** [comment]
> It seems to be related to remote builds.

Hmm, I don't use remote builds though

> I have to `sudo kill -9` the processes

Which processes? `killall -9 nix-daemon` did kill/restart them, but didn't solve the problem for me.

**Comment by 06kellyjac** [comment]
if you run that as a normal user it'll ignore you

**Comment by kenranunderscore** [comment]
I ran it as `sudo` and it didn't fix my problem either. Also no remote builds here.
*Even a reboot didn't fix it at first.*

After the reboot and failed attempt I tried to `nix-collect-garbage -d && sudo nix-collect-garbage -d`, after which it started working again.

**Comment by raboof** [comment]
> if you run that as a normal user it'll ignore you

(I sudo'ed, and verified in `ps aux` that the processes were really gone/replaced)

**Comment by 06kellyjac** [comment]
ohh, **did** kill them. I misread.

Well after I ran into this problem a couple times and removed remote builders I've not encountered it again since.
I did also notice during those period that ping-ing cache.nixos.org also hung for like 10 seconds plus

**Comment by Artturin** [comment]
I can repro

@edolstra @roberth

**Comment by roberth** [comment]
@Artturin 

> I can repro

How?

Here's a fix for a deadlock involving CurlDownloader and the interrupt handlers. May or may not be related. Does it repro with this fix applied?
 - https://github.com/NixOS/nix/pull/6052

**Comment by Artturin** [comment]
> @Artturin
> 
> > I can repro
> 
> How?
> 
> Here's a fix for a deadlock involving CurlDownloader and the interrupt handlers. May or may not be related. Does it repro with this fix applied?
> 
>     * [Fix deadlocked nix-daemon zombies on darwin #3294 nix#6052](https://github.com/NixOS/nix/pull/6052)

Sorry I should have said that I have the same issue.

I'll apply that pr and see if I get this issue anymore

**Comment by Artturin** [comment]
i've been running nix with this overlay since 5 days ago and i haven't gotten the same problem
```
      (self: super: {
        nixUnstable = super.nixUnstable.overrideAttrs (old: {
          src = super.fetchFromGitHub {
            owner = "nixos";
            repo = "nix";
            rev = "a768e85e2fb3b0500829bc42cdc137176481bedf";
            sha256 = "sha256-XtjveEPLGhHPh95+yequcLajtyCskSUTx+XZsP4UqA0=";
          };
          patches = (old.patches or [ ]) ++ [
            (super.fetchpatch {
              url = "https://github.com/NixOS/nix/commit/c3b942e0fc4777f9033f614b6b1f462c0f8c473e.patch";
              sha256 = "sha256-LQ5zkXwv1/3DoZz6CevyipdHTdxX7XlXv3nbmgGsngA=";
            })
          ];
        });
      })
```

updated https://github.com/NixOS/nixpkgs/pull/158455 to the newest nix commit to include the pr

**Comment by miallo** [comment]
@Artturin Just as a warning that you might just be lucky: I also ran into the stuck downloads once and after following the instructions by @kenranunderscore (`nix-collect-garbage`) I could do the updates and for the last 7 days it also worked for me.

Did anyone figure out how to reproduce this bug?

**Comment by raboof** [comment]
I also haven't seen stuck downloads for the past week or so...

**Comment by r-burns** [comment]
I'm still getting this on NixOS unstable with Nix 2.6.1. Is there a workaround that doesn't require collecting garbage? I have a couple compute-expensive derivations on my hard drive I'd rather not have to rebuild....

**Comment by raboof** [comment]
I haven't seen this for a while now. Let's close for now and re-open when we see evidence it's still a problem.

**Comment by sheldonneuberger-sc** [comment]
I've hit this issue a few times in the past week. It's always the first package that is attempted to be copied from cache.nixos.org, which for me is `/nix/store/8pil82i9qfww8l752158kqsa8dc2fzfl-python3.9-pyyaml-6.0`. Haven't found any more info on it yet.

 - system: `0`
 - host os: `Darwin 21.5.0, macOS 10.16`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.10.3`

**Comment by raboof** [comment]
Interesting. The problem hasn't returned for me, and I think in my case it wasn't the first package, so it might not be the same problem. Perhaps it would be better to add a new issue (and link to this one for possible background) rather than re-open this?

**Comment by uuuvn** [comment]
I was hit by a similar problem (but copying files from host to a vds via remote nixos-rebuild). nix-collect-garbage on both host and vds did nothing but upgrading host's system did solved it

**Comment by kquick** [comment]
Encountered a similar issue: nix operations would just hang when trying to download (the perl package in my case).  Was able to resolve with: 
```
$ systemctl stop nix-daemon.socket
$ systemctl stop nix-daemon
$ nix store gc
$ systemctl start nix-daemon.socket
$ systemctl start nix-daemon
```
plus killing all active/hung `nix` attempts.

**Comment by nebez** [comment]
I've been having this issue in my Docker containers. `FROM nixos/nix:latest` has been failing for ~a few weeks now on a fresh machine with no docker cache, didn't think much of it and relied on the docker cache.

After losing the docker cache, I have been unable to rebuild my images at all. Stuck at `copying path '/nix/store/[...]' from 'https://cache.nixos.org'...` indefinitely.

Pinning my tags to `FROM nixos/nix:2.20.0` fixed it. Will debug more later.

**Comment by Shados** [comment]
Saw this on a system today. I *think* in this particular instance it was triggered through some odd interaction with one remote builder being unavailable; removing it from the builders list, `kill -9`'ing the stuck nix-daemon processes and restarting the daemon overall seemed to fix the issue. Kinda surprised that `systemctl stop nix-daemon.service` reported success despite there being zombie `nix-daemon` processes still running -- shouldn't systemd have been able to pick up on those through its cgroup shenanigans?

**Comment by ocharles** [comment]
I just ran into this. The fix for me was to kill all stuck `nix-store` processes, then `systemctl restart nix-daemon`.

**Comment by gador** [comment]
Just had this today. It seemed to be a stuck `nix-store` process as well. In my case by `hdyra`. Reboot hadn't solved the issue, probably because `hydra` gut stuck instantly after reboot again. Stopping my `hydra` instance killed the process and my normal build continued

**Comment by shazow** [comment]
I just ran into this because I ctrl-z suspended a `home-manager switch --flake .` for about an hour, then tried resuming and it would persistently get stuck at the last package it was processing (even if I aborted and tried again). It kept getting stuck until I rebooted. I suspect killing a rogue process might have helped too, but I didn't check for it first.

**Comment by hedefalk** [comment]
Happened to me today. Stuck at:

```
sudo nix run nix-darwin/nix-darwin-25.05#darwin-rebuild -- switch --flake path:$HOME/dotfiles/#$(scutil --get LocalHostName)
warning: $HOME ('/Users/viktor') is not owned by you, falling back to the one defined in the 'passwd' file ('/var/root')
building the system configuration...
error: interrupted by the user
```

Thought it was the café wifi network so went home. Then same at home so `sudo pkill -9 nix-daemon` and then all was fine.

**Comment by hedefalk** [comment]
>I just ran into this because I ctrl-z suspended a home-manager switch --flake .

I think this happened for me because I'm using an ai agent to help me with my nix setup within zed and it wants to verify directly in the agent flow with a `nix-darwin rebuild --switch` that times out.

**Comment by mawildoer** [comment]
Hitting this repeatedly and I thiiiink what's happening, at least in my case is a deadlock between the `nix copy` client and nix daemon, along roughly these lines

During nix copy, the flow requires both sides to send and receive on the same socket:
1. Client sends NAR data (large, potentially hundreds of MB) to daemon
2. Daemon needs to send back acknowledgments/responses while receiving

```
  Client                          Unix Socket                         Daemon
     |                                                                   |
     |--- write(NAR chunk) -----> [send buffer FULL] -------> [recv buffer FULL]
     |    BLOCKED in                                           BLOCKED in
     |    sock_alloc_send_pskb                                 unix_stream_read_generic
     |                                                                   |
     |    Can't read daemon's     <----- [response data] ----  Can't send
     |    response because                                     response because
     |    blocked on write                                     blocked on read
```
- Client is blocked trying to write more NAR data, but the socket send buffer is full (214KB queued)
- Daemon is blocked trying to read, but it's expecting a protocol header/command, not the raw NAR data already sitting in its 197KB recv buffer
- Neither side can make progress because each is waiting for the other to drain data first

Certainly not sure here, but in my case I've been hitting ~80 zombie copy processes that seem to log-jam. Killing them will almost instantly fix it
